README
= LINZ Geodetic modules

The Geodetic modules provide tools for converting between different coordinate
systems in use in New Zealand.  The coordinate systems are loaded a coordinate
system definition file, by default installed at a location defined by the 
COORDSYSDEF environment variable.

Copyright 2014 Crown copyright (c) Land Information New Zealand and the New
Zealand Government.

== Dependencies

 * Perl 5.88 or greater, plus
    - Time::JulianDay

== Compatibility

Tested on Windows XP and Server 2003, and Ubuntu 10.4 (Lucid)

== Install

This is a pure perl implementation of coordinate conversions.  Installation
uses the Module::Build process listed below. 

-- Simple install

    perl Build.PL
    ./Build install	


debian/changelog
liblinz-geodetic-perl (2.17.2-1) xenial; urgency=medium

  * fix: Error in time interpolation in version 1 format models

 -- Chris Crook <ccrook@linz.govt.nz>  Tue, 16 Aug 2022 09:00:00 +1300

liblinz-geodetic-perl (2.17.1-1) xenial; urgency=medium

  * fix: Error reading local vertical datum offsets

 -- Chris Crook <ccrook@linz.govt.nz>  Fri, 12 Aug 2022 15:00:00 +1300

liblinz-geodetic-perl (2.17.0-1) xenial; urgency=medium

  * feat: Add missing functionality to list vertical datum codes

 -- Chris Crook <ccrook@linz.govt.nz>  Thu, 21 Jul 2022 15:00:00 +1300

liblinz-geodetic-perl (2.16.1-1) xenial; urgency=medium

  * fix: Kermadec Island map sheet extents

 -- Chris Crook <ccrook@linz.govt.nz>  Wed, 22 Jun 2022 15:00:00 +1300

liblinz-geodetic-perl (2.16.0-1) xenial; urgency=medium

  * Adding Kermadec Island map sheet references

 -- Chris Crook <ccrook@linz.govt.nz>  Wed, 22 Jun 2022 15:00:00 +1300

liblinz-geodetic-perl (2.15.1-2) xenial; urgency=medium

  * Fixing packaging for bionic

 -- Chris Crook <ccrook@linz.govt.nz>  Thu, 28 Mar 2022 09:00:00 +1300

liblinz-geodetic-perl (2.15.1-1) xenial; urgency=medium

  * Fixing error message

 -- Chris Crook <ccrook@linz.govt.nz>  Tue, 19 Mar 2019 12:55:23 +1300

liblinz-geodetic-perl (2.15.0-1) xenial; urgency=medium

  * Adding code for NZTopo250MapRef

 -- Chris Crook <ccrook@linz.govt.nz>  Tue, 19 Mar 2019 09:44:48 +1300

liblinz-geodetic-perl (2.14.0-1) xenial; urgency=medium

  * Adding MGRS map reference code

 -- Chris Crook <ccrook@linz.govt.nz>  Thu, 13 Sep 2018 05:56:24 +1200

liblinz-geodetic-perl (2.13.1-1) xenial; urgency=medium

  * Fixing bug in setting current version

 -- Chris Crook <ccrook@linz.govt.nz>  Thu, 17 May 2018 10:34:25 +1200

liblinz-geodetic-perl (2.13.0-1) xenial; urgency=medium

  * Addng support for version 3 of LINZ deformation model format

 -- Chris Crook <ccrook@linz.govt.nz>  Thu, 17 May 2018 07:23:22 +1200

liblinz-geodetic-perl (2.12.0-1) xenial; urgency=medium

  * Adding metres_per_degree to Ellipsoid
  * Making ordinate orders consistent
  * Fixing bug with multiple datum transformation and deformation model not in first or last step
  * Adding some rudimentary tests

 -- Chris Crook <ccrook@linz.govt.nz>  Wed, 20 Dec 2017 19:45:41 +1300

liblinz-geodetic-perl (2.11.0-1) xenial; urgency=medium

  * Adding support for coordinate system aliases

 -- Chris Crook <ccrook@linz.govt.nz>  Fri, 09 Jun 2017 10:43:42 +1200

liblinz-geodetic-perl (2.9.0-1) trusty; urgency=medium

  * Updating MANIFEST
  * Adding optional "offset" keyword to height reference surface definition

 -- Chris Crook <ccrook@linz.govt.nz>  Tue, 25 Oct 2016 11:50:50 +1300

liblinz-geodetic-perl (2.8.1-1) trusty; urgency=medium

  * Fixing bug handling height reference surfaces when not in geoid coordinate system

 -- Chris Crook <ccrook@linz.govt.nz>  Fri, 26 Aug 2016 17:03:59 +1200

liblinz-geodetic-perl (2.8.0-1) trusty; urgency=medium

  * Refactoring height reference systems
  * Bug fix in getting height reference system name

 -- Chris Crook <ccrook@linz.govt.nz>  Wed, 15 Jun 2016 15:30:38 +1200

liblinz-geodetic-perl (2.7.0-1) trusty; urgency=medium

  * Adding simplistic grid format conversion

 -- Chris Crook <ccrook@linz.govt.nz>  Thu, 21 Apr 2016 15:32:48 +1200

liblinz-geodetic-perl (2.6.1-1) trusty; urgency=medium

  * Bug fix

 -- Chris Crook <ccrook@linz.govt.nz>  Thu, 24 Sep 2015 06:03:18 +1200

liblinz-geodetic-perl (2.6.0-1) trusty; urgency=medium

  * Improving access to projection information

 -- Chris Crook <ccrook@linz.govt.nz>  Sat, 19 Sep 2015 09:07:10 +1200

liblinz-geodetic-perl (2.5.1-1) trusty; urgency=medium

  * Updating to support pre 5.11 perl

 -- Chris Crook <ccrook@linz.govt.nz>  Thu, 10 Sep 2015 11:36:57 +1200

liblinz-geodetic-perl (2.5.0-1) trusty; urgency=medium

  * Added IERS format for reference frame definition

 -- Chris Crook <ccrook@linz.govt.nz>  Fri, 01 May 2015 14:34:02 +1200

liblinz-geodetic-perl (2.4.0-1) trusty; urgency=medium

  * Version 2 binary format deformation model - WIP
  * Better debugging output
  * Fixing error in handling dimension of patch deformation

 -- Chris Crook <ccrook@linz.govt.nz>  Thu, 01 Jan 2015 07:08:19 +1300

liblinz-geodetic-perl (2.3.0-1) precise; urgency=low

  * Adding Polar Stereographic projection
  * Adding functions for testing whether need epoch, and improving handling of epoch in conversions

 -- Chris Crook <ccrook@linz.govt.nz>  Fri, 01 Aug 2014 09:39:32 +1200

liblinz-geodetic-perl (2.2.0-1) precise; urgency=low

  * Adding selection of meridional circuits based on lon/lat

 -- Chris Crook <ccrook@linz.govt.nz>  Mon, 09 Jun 2014 14:02:34 +1200

liblinz-geodetic-perl (2.1.0-1) precise; urgency=low

  * Added reporting of deformation model version

 -- Chris Crook <ccrook@linz.govt.nz>  Thu, 29 May 2014 20:16:00 +1200

liblinz-geodetic-perl (2.0-1) stable; urgency=low

  * Initial release 

 -- Chris Crook <ccrook@linz.govt.nz>  Thu, 01 May 2014 11:51:14 +1200


tools/build_circuit_file.pl
use strict;

my $inputfile='merid_buff.dump';
my $outputfile='nz_merid_circuit.dat';

open( my $if, "<$inputfile") || die "Cannot open $inputfile\n";


my %crdsys=(
    2=>'AMURTM2000',
    5=>'PLENTM2000',
    8=>'BLUFTM2000',
    11=>'BULLTM2000',
    15=>'CHATTM2000',
    18=>'COLLTM2000',
    21=>'GAWLTM2000',
    24=>'GREYTM2000',
    28=>'HAWKTM2000',
    30=>'HOKITM2000',
    33=>'JACKTM2000',
    36=>'KARATM2000',
    39=>'LINDTM2000',
    42=>'MARLTM2000',
    45=>'EDENTM2000',
    48=>'PLEATM2000',
    51=>'YORKTM2000',
    54=>'NICHTM2000',
    57=>'NELSTM2000',
    62=>'TAIETM2000',
    65=>'OBSETM2000',
    68=>'OKARTM2000',
    86=>'POVETM2000',
    91=>'TARATM2000',
    94=>'TIMATM2000',
    97=>'TUHITM2000',
    100=>'WAIRTM2000',
    103=>'WANGTM2000',
    106=>'WELLTM2000',
);

my @circuits=();
while( my $line=<$if> )
{
    next if $line !~ /^OGRFeature\(/;
    my $cosid=0;
    my $crds='';
    my $name='';
    while( $line = <$if> )
    {
        if( $line=~ /^\s*cos_id\s+\(.*\)\s*\=\s*(\d+)(\.0*)?\s*$/ )
        {
            $cosid=$1;
        }
        elsif( $line=~ /^\s*name\s+\(.*\)\s*\=\s*(.*?)\s*$/ )
        {
            $name=$1;
        }
        elsif( $line=~ /^\s*POLYGON\s*\(/ )
        {
            $crds=$';
            last;
        }
        elsif( $line =~ /^\s*$/ )
        {
            last;
        }
    }
    if( $cosid && $name && $crds )
    {
        if( ! exists($crdsys{$cosid}))
        {
            print "Don't know circuit $cosid\n";
            next;
        }
        my @ccrds=();
        my ($lnmin,$lnmax,$ltmin,$ltmax);
        while($crds=~ /[\(\,]([\d\.\-]+)\s+([\d\.\-]+)/g )
        {
            my $ln=sprintf("%.4f",$1);
            my $lt=sprintf("%.4f",$2);
            if( ! @ccrds ){ $lnmin=$lnmax=$ln; $ltmin=$ltmax=$lt; }
            if( $ln < $lnmin ){ $lnmin=$ln; } elsif ($ln > $lnmax) { $lnmax=$ln; }
            if( $lt < $ltmin ){ $ltmin=$lt; } elsif ($lt > $ltmax) { $ltmax=$lt; }
            push(@ccrds,[$ln,$lt]);
        }
        my $ncrd=scalar(@ccrds);
        print "$crdsys{$cosid} $lnmin $lnmax $ltmin $ltmax $ncrd\n";
        foreach my $x (@ccrds)
        {
            print $x->[0]," ",$x->[1],"\n";
        }
    }
}
